File: src/main/kotlin/org/example/schwimmen/eingabe/MinMax.kt
package org.example.schwimmen.eingabe

fun parseMinMax(data: String): Map<String, MinMax> = parseMinMax(data.lines().map { it.split("\t") })

fun parseMinMax(rows: List<List<String>>): Map<String, MinMax> =
    rows
        .filter { it.size >= 2 }
        .associate { it[0] to MinMax(min = it.getOrNull(1)?.toIntOrNull(), max = it.getOrNull(2)?.toIntOrNull()) }

data class MinMax(
    val min: Int?,
    val max: Int?,
)

File: web/src/lib/schwimmen/eingabe/minmax.ts
export interface MinMax {
  min: number | undefined;
  max: number | undefined;
}

export function parseMinMax(data: string): Map<string, MinMax> {
  return parseMinMaxFromGrid(data.split("\n").map((row) => row.split("\t")));
}

function maybeNumber(cell: string | undefined): number | undefined {
  if (cell !== undefined && cell.trim().length != 0) {
    return Number(cell);
  } else {
    return undefined;
  }
}

export function parseMinMaxFromGrid(rows: string[][]): Map<string, MinMax> {
  return new Map<string, MinMax>(
    rows

      .filter((row) => row.length >= 2)
      .map((row) => [row[0], { min: maybeNumber(row[1]), max: maybeNumber(row[2]) }]),
  );
}
